rust/arrow/src/array/array_primitive.rs

```diff
@@ -86,13 +86,9 @@ impl<T: ArrowPrimitiveType> PrimitiveArray<T> {
     }
 
     /// Returns the primitive value at index `i`.
-    ///
-    /// Note this doesn't do any bound checking, for performance reason.
-    /// # Safety
-    /// caller must ensure that the passed in offset is less than the array len()
+    #[inline]
     pub fn value(&self, i: usize) -> T::Native {
-        let offset = i + self.offset();
-        unsafe { *self.raw_values.as_ptr().add(offset) }
+        self.values()[i]
```

Comment:
Thinking about it, I am not sure we want this: this will have severe implications to almost all uses of `value()`. Could you run the benches before and after this PR?

My suggestion is that we mark this method as `unsafe`, so that it is clear that people need to be conscious about it. Since this will impact a large part of our code base (and likely others), I suggest that we do this change on a separate PR

Reply:
I agree, that we want to guide API users to switch to values().  My hesitancy with making/leaving it unsafe is that I think it's the only Array value(x) method when combined with the other changes in this PR (which is why it got added as a followon commit).  That leads to awkwardness in macros, such as the comparison kernel, where it effectively calls "$left.value(x)",  it could be a papercut for macro-style invocation.  

(There may be an argument that a trait could abstract away the need for macros altogether... but my experiments to rework the comparison kernel to that effect are incomplete - and partly sidelined with various other experiments I'm toying with)

I would expect that in the exact for loop you mention, the compiler could recognize the satisfied loop bound since both len() and .value(i) are inlinable and use the same underlying field, but I expect it is also easy to complicate on that loop to defeat that by having a less direct relationship between i & array.len().  The pretty printing logic probably would do additional bounds checks now, as it iterates based on  len() from the record batch, which is semantically equal, but nothing in the code enforces that in a way I would expect the compiler to understand.

I'll launch some benchmarks of head, master, and the commit just before the PrimitiveArray safety change.

Reply:
I think that we may need to park this change until we migrate our code base to use `values()` whenever possible. 

If we merge this one, we get a major performance degradation. If we add `unsafe`, we need to add `unsafe` in a lot of places. Neither are great options :)

Reply:
That makes sense.  I'll rebase dropping that commit.

Reply:
safety overhead on a few benchmarks seems to be 15-30%
```
group                           master-                                safe-prim-                             unsafe-prim-
-----                           -------                                ----------                             ------------
add 512                         1.00    307.4±4.79ns        ? B/sec    1.07    327.7±6.48ns        ? B/sec    1.30    400.5±7.09ns        ? B/sec
add_nulls_512                   1.20    447.8±8.09ns        ? B/sec    1.32    490.9±3.41ns        ? B/sec    1.00    372.2±5.96ns        ? B/sec
buffer_bit_ops or               1.17   731.0±20.02ns        ? B/sec    1.19    743.7±5.43ns        ? B/sec    1.00   624.0±10.86ns        ? B/sec
cast date64 to date32 512       1.00      7.4±0.22µs        ? B/sec    1.10      8.2±0.07µs        ? B/sec    1.03      7.7±0.22µs        ? B/sec
cast i64 to string 512          1.00     24.8±0.33µs        ? B/sec    1.09     27.1±0.79µs        ? B/sec    1.10     27.3±0.41µs        ? B/sec
cast time32s to time32ms 512    1.00   917.0±19.20ns        ? B/sec    1.21  1113.0±22.04ns        ? B/sec    1.07    984.5±7.53ns        ? B/sec
equal_512                       1.00     44.1±0.81ns        ? B/sec    1.13     49.9±0.65ns        ? B/sec    1.01     44.4±0.27ns        ? B/sec
min nulls string 512            1.04      6.4±0.12µs        ? B/sec    1.14      7.0±0.15µs        ? B/sec    1.00      6.2±0.09µs        ? B/sec
multiply 512                    1.14   471.5±10.15ns        ? B/sec    1.18    487.0±4.72ns        ? B/sec    1.00    413.4±6.16ns        ? B/sec
not                             1.00  1102.3±14.55ns        ? B/sec    1.12  1235.9±19.91ns        ? B/sec    1.02  1120.3±19.83ns        ? B/sec
```

Reply:
I think that I covered all of those in #9291 . Good that there is no impact on `sort`, which was the only case where I could not optimize (i.e. we really needed the bound check).

The larger task atm is to convert all unit-tests that use `value(i)` to `values()[i]`.

Fyi, the reason I propose keeping `value(i)` as `unsafe` instead of un-performant is that there is no way to efficiently access the value `i` without knowing the internals of the primitiveArray (i.e. pointer and offset). imo this method does serve the purpose of abstracting out those details. It just happens to be `unsafe` and we currently not marking it as such. However, please say if you disagree here.

wrt to the `comparison kernel`, we can create a generic for primitives alone for that use-case, no?

Let me know what do you think

Reply:
#9291 is good progress towards eliminating use of the function.

And certainly, we could 'split' the macro for primitives as a quick fix to get rid of the call to the function.  I've been experimenting with an alternative approach that might be a bit more flexible to multiple use cases, described at the bottom of this comment.

I am quite torn about whether I think value should or should not be in the interface.
# Reasons to drop value(i) -> T::Native
I think that even if value(i) was dropped from the PrimitiveArray impl's, efficient random access to items without a bounds check can still be achieved through ```unsafe{*primitive_array.values().get_unchecked(i)}```  (the extra * because get_unchecked() returns a ref to the value).

I'm not sure I have any example code or measurements to demonstrate it on hand, but I am certain I saw the silently-unsafe implementation```x.values().iter().zip(y.values().iter())``` did (slightly) outperform ```(0..x.len()).map(|i|{x.value(i),y.value(i)}```.  I believe it was when I was playing with non-simd arithmetic kernels....  So that is the root of my hesitancy, is I'm worried it doesn't actually escape any overhead, and unintentionally lead people away from a more reliable/performant way.  IF there is a context where unsafe{x.value(i)} beats the performance of unsafe{*x.values().get_unchecked(i)}

# Reasons to keep value(i) -> T::Native
All other array implementations have value functions as far as I recall, so it is a nice 'consistency'.  

In the back of my mind, the biggest argument to keep value(i) is for api consistency... so long term, a 'trait' may be the place where it might fit best?  Very roughly, I'm thinking:
```
trait TypedArrowArray : ArrowArray {
   type RefType;
   fn is_valid(i:usize) -> bool; //bounds check
   unsafe fn is_valid(i:usize) -> bool; //no bounds check
   fn value(i:usize) -> RefType;  //bounds check
   unsafe fn value_unchecked(i:usize) -> RefType; //no bounds checked
   fn iter() -> impl Iterator<Option<RefType>>;
   fn iter_values() -> impl Iterator<RefType>;
}
impl <T: ArrowPrimitiveType> TypedArrowArray<&T::Native> for PrimitiveArray<T> { ... }
impl TypedArrowArray<ArrayRef> for GenericListArray<T> { ... }
//and similar for string/binary. ... I am not sure whether struct arrays could fit... Dictionary would not give access to 'keys', only to the values referenced by each key?  Union would require some kind of RefType that can downcast into the actual value?
```
Of course, I am uncertain how much overhead the 'standarization' such a trait impl implies would bring...  would any kernels actually benefit from using generic implementations against such an api, or will they always go down to the concrete type to squeeze little short-cuts out that don't fit in the generic interface?  I'm unsure, so very (very, very) slowly experimenting...

# Summary

So in short, my thoughts are:
* I think that leaving value(i) safety consideration out of this PR makes sense.  I've rebased to drop that out - although I did leave the additional values() test code.
* Marking it unsafe in the near future is absolutely better than being silently-unsafe.  The argument that adding bounds-checks could silently impact external users is reasonable, taking unsafe has the larger 'warning' so that the change isn't missed.
* Longer term, the options of deprecating it, or explicitly moving it into an trait impl are both contenders in my mind... but neither option is directly relevant to this PR.

Let me know if that seems reasonable.

Reply:
Just looking over the benchmarks again, one of the biggest changes is actually a performance improvement while doing safe access... which makes very little sense... so I think I'll run the benchmarks again overnight.